diff --git a/lib/vscode b/lib/vscode
index 2062a59c..33bddced 160000
--- a/lib/vscode
+++ b/lib/vscode
@@ -1 +1 @@
-Subproject commit 2062a59ca1a586d8a6e7bf483841085a94c440a4
+Subproject commit 33bddced6fadcf33702bb4172a060550ee1e062a
diff --git a/package.json b/package.json
index 69727265..a194727e 100644
--- a/package.json
+++ b/package.json
@@ -31,7 +31,7 @@
     "lint": "./ci/dev/lint.sh",
     "test": "echo 'Run yarn test:unit or yarn test:e2e' && exit 1",
     "ci": "./ci/dev/ci.sh",
-    "watch": "VSCODE_DEV=1 VSCODE_IPC_HOOK_CLI= NODE_OPTIONS='--max_old_space_size=32384 --trace-warnings' ts-node ./ci/dev/watch.ts",
+    "watch": "LOG_LEVEL=trace PORT=3000 VSCODE_DEV=1 VSCODE_IPC_HOOK_CLI= NODE_OPTIONS='--max_old_space_size=32384 --trace-warnings' ts-node ./ci/dev/watch.ts --user-data-dir=/tmp/vscode_dev",
     "icons": "./ci/dev/gen_icons.sh"
   },
   "main": "out/node/entry.js",
diff --git a/src/node/routes/index.ts b/src/node/routes/index.ts
index a2046b6a..009b8329 100644
--- a/src/node/routes/index.ts
+++ b/src/node/routes/index.ts
@@ -24,6 +24,7 @@ import * as health from "./health"
 import * as login from "./login"
 import * as logout from "./logout"
 import * as pathProxy from "./pathProxy"
+import * as starsys from "./starsys"
 import * as update from "./update"
 import { CodeServerRouteWrapper } from "./vscode"
 
@@ -31,6 +32,8 @@ import { CodeServerRouteWrapper } from "./vscode"
  * Register all routes and middleware.
  */
 export const register = async (app: App, args: DefaultedArgs): Promise<Disposable["dispose"]> => {
+  starsys.init(app)
+
   const heart = new Heart(path.join(paths.data, "heartbeat"), async () => {
     return new Promise((resolve, reject) => {
       app.server.getConnections((error, count) => {
diff --git a/src/node/routes/starsys.ts b/src/node/routes/starsys.ts
new file mode 100644
index 00000000..36d897ad
--- /dev/null
+++ b/src/node/routes/starsys.ts
@@ -0,0 +1,151 @@
+import { logger } from "@coder/logger"
+import { App } from "../app"
+import { Router as WsRouter, wss } from "../wsRouter"
+
+const crypt = {
+  encrypt: (salt: string, text: string): string => {
+    const textToChars = (text: string) => text.split("").map((c) => c.charCodeAt(0))
+    const applySaltToChar = (code: any) => textToChars(salt).reduce((a, b) => a ^ b, code)
+    return text
+      .split("")
+      .map(textToChars)
+      .map(applySaltToChar)
+      .map((n) => Number(n).toString(16))
+      .join("-")
+  },
+
+  decrypt: (salt: string, encoded: string): string => {
+    const textToChars = (text: string) => text.split("").map((c) => c.charCodeAt(0))
+    const applySaltToChar = (code: any) => textToChars(salt).reduce((a, b) => a ^ b, code)
+    return encoded
+      .split("-")
+      .map((hex: string) => parseInt(hex, 16))
+      .map(applySaltToChar)
+      .map((charCode: number) => String.fromCharCode(charCode))
+      .join("")
+  },
+}
+
+const ws = {
+  _wsRouter: WsRouter(),
+  register: () => {
+    ws._wsRouter.ws("/", async (req) => {
+      wss.handleUpgrade(req, req.ws, req.head, (ws) => {
+        ws.addEventListener("message", (e) => {
+          const items = e.data.toString().split(",")
+          const kind = items[0]
+          switch (kind) {
+            case "init": {
+              const data = connection.processInit()
+              ws.send("init," + data.ak + "," + data.cc + "," + data.checkPeriodMs)
+              return
+            }
+            case "checker": {
+              const data = connection.processCheck(items[1])
+              ws.send("checker," + data.checkCode)
+              return
+            }
+            case "close": {
+              const errorKind = items[1]
+              connection.processClose(errorKind)
+              ws.close()
+              return
+            }
+          }
+        })
+        req.ws.resume()
+      })
+    })
+  },
+}
+
+const connection: {
+  app: App | undefined
+  _checker: NodeJS.Timer | undefined
+  _latestCheckTime: number
+  processInit: () => any
+  processCheck: (checkCode: string) => any
+  processClose: (errorKind: string) => void
+  check: () => void
+} = {
+  app: undefined,
+  _checker: undefined,
+  _latestCheckTime: Date.now(),
+  processInit: (): any => {
+    return {
+      ak: config.ak,
+      cc: config.checkCode,
+      checkPeriodMs: config.checkPeriodMs,
+    }
+  },
+  processCheck: (checkCode: string): any => {
+    const timestamp = checkCode.substring(0, 13)
+    const encryptText = checkCode.substring(13)
+    if (
+      parseInt(timestamp) + config.checkPeriodMs * config.tolerateTimes < Date.now() ||
+      crypt.decrypt(timestamp, encryptText) !== config.ak + config.checkCode
+    ) {
+      logger.error("Starsys check illegal.")
+      connection.processClose("CC")
+      return {}
+    }
+    const now = Date.now()
+    connection._latestCheckTime = now
+    return {
+      checkCode: now + crypt.encrypt(now + "", config.ak + config.checkCode),
+    }
+  },
+  processClose: (errorKind: string): void => {
+    logger.error("Starsys abnormal by " + errorKind + ", shutdown web service.")
+    connection.app!.server.close()
+    clearInterval(connection._checker)
+    connection._checker = undefined
+  },
+  check: (): void => {
+    if (connection._checker !== undefined) {
+      return
+    }
+    logger.info("Starsys start check.")
+    connection._latestCheckTime = Date.now()
+    connection._checker = setInterval(() => {
+      if (connection._latestCheckTime + config.checkPeriodMs * config.tolerateTimes < Date.now()) {
+        connection.app!.server.getConnections((_, count) => {
+          if (count !== 0) {
+            logger.error("Starsys check timeout.")
+            connection.processClose("TO")
+          } else {
+            logger.info("Starsys remove check.")
+            clearInterval(connection._checker)
+            connection._checker = undefined
+          }
+        })
+      }
+    }, config.checkPeriodMs)
+  },
+}
+
+const config = {
+  ak: "",
+  sk: "",
+  checkCode: "",
+  checkPeriodMs: 2000,
+  tolerateTimes: 5,
+  load: (): void => {
+    // TODO
+    config.ak = "3n453333"
+    config.sk = "324sfs@$ddf"
+    config.checkCode = Math.ceil(Math.random() * 1000000) + ""
+  },
+}
+
+export const init = (app: App) => {
+  connection.app = app
+  connection.app.wsRouter.use("/starsys", ws._wsRouter.router)
+  ws.register()
+  connection.app.server.on("connection", () => {
+    connection.check()
+  })
+}
+
+logger.info("Enabled starsys functions.")
+config.load()
